### PR TITLE
FH-3606 - Add command fhc projects to V3 standard and add unit test

### PR DESCRIPTION
**Pull request by camilamacedo86**
JIRA: https://issues.jboss.org/browse/FH-3606

Changes:
* Add the V3 standard
* Divide as category
* Improve output create, update and delete
* Add unit tests

Automated Test affect. 

Important highlight: It is one of the advantages of we do this changes, solve the technical debts, and make all work with the same standard. Currently, has more than one way that we need call the methods of the fhc-tool when it is used as dependence.

JIRA: https://issues.jboss.org/browse/FH-3670

Following local test 

* To list ( `./bin/fhc.js projects list` )

screen shot 2017-06-01 at 6 00 04 pm

* To list with the json format. ( `./bin/fhc.js projects list --json` )

screen shot 2017-06-02 at 10 41 29 am

* To list ( `./bin/fhc.js projects list --autherEmail` )

````
Camilas-MBP:fh-fhc cmacedo$ ./bin/fhc.js projects list --authorEmail=" "
fhc ERR! Not found projects with the authorEmail ' ' .
fhc Command executed with error
Camilas-MBP:fh-fhc cmacedo$ 
````
screen shot 2017-06-01 at 6 05 38 pm

* To update: 

`````
Camilas-MBP:fh-fhc cmacedo$ ./bin/fhc.js projects update ybmcmv3fuola27sxqxknzgp7 propName propValue
Project 'TestAndroid' update successfully with the propName 'propName' and propValue 'propValue'
Camilas-MBP:fh-fhc cmacedo$ 
``````

* To delete:

````
Camilas-MBP:fh-fhc cmacedo$ ./bin/fhc.js projects delete ll3cuwvxomdxcdedbgu6pswy
Project 'TestFHC' delete successfully
Camilas-MBP:fh-fhc cmacedo$ ./bin/fhc.js projects delete ll3cuwvxomdxcdedbgu6pswy
fhc ERR! Error: Error - not 2xx status code. (404)
fhc ERR! {"message":"Cannot find entity for guid=ll3cuwvxomdxcdedbgu6pswy","status":"error","result":null}
fhc ERR! Request ID:  01c188e1-f88d-4250-b311-82de7d050ca1
fhc ERR! 
fhc ERR! System Darwin 15.6.0
fhc ERR! command "/Users/cmacedo/.nvm/versions/node/v4.4.3/bin/node" "/Users/cmacedo/work/fh-fhc/bin/fhc.js" "projects" "delete" "ll3cuwvxomdxcdedbgu6pswy"
fhc Command executed with error
Camilas-MBP:fh-fhc cmacedo$ 
````

* To clone:

screen shot 2017-06-01 at 6 13 13 pm

### CODE REVIEW

File: lib/cmd/fh3/projects/clone.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  'demand' : ['project'],
+  'alias' : {
+    'project': 'p',
+    0 : 'project'
+  },
+  'describe' : {
+    'project' : i18n._('Unique 24 character GUID of the project')
+  },
+  'customCmd': function(params, cb) {
+    common.listApps(params.project, function(err, apps) {
```
Comment (wtrocki) [inline lib/cmd/fh3/projects/clone.js]: FYI - common commands can be moved to fh3 format as well

File: lib/cmd/fh3/projects/create.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    cmd : 'fhc projects create --title=<title> --template=<template>',
+    desc : i18n._('Create a project with the <title> and <template>')
+  },{
+    cmd : 'fhc projects create --title=<title> --template=<template> --env=<environment>',
+    desc : i18n._('Create a project with the <title> and <template>, after deploy it on <environment>')
+  }],
+  'demand' : ['title'],
+  'alias' : {
+    'title' : 't',
+    'template' : 't',
```
Comment (wei-lee) [inline lib/cmd/fh3/projects/create.js]: not sure if same alias works?

File: lib/cmd/fh3/projects/create.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+var ini = require('../../../utils/ini');
+var templates = require('../../common/templates.js');
+var util = require('util');
+
+
+module.exports = {
+  'desc' : i18n._('Create project'),
+  'examples' :
+  [{
+    cmd : 'fhc projects create --title=<title>',
```
Comment (wtrocki) [inline lib/cmd/fh3/projects/create.js]: That's breaking change. Original command was:
`fhc projects create name` 
Title is "position parameter" to maintain backwards compatibility.

If we want to change it  then it's best to notify and update documentation as well.

**Comment by wtrocki** [comment]
Verified changes locally. Everything works fine.

File: lib/cmd/fh3/projects/create.js (hunk 2 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/projects/create.js]: @wei-lee Tks for the review and to finding it. It doesn't will work and will fix it.

File: lib/cmd/fh3/projects/clone.js (hunk 1 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/projects/clone.js]: Tks for the review 👍  Done this change :-)

File: lib/cmd/fh3/projects/create.js (hunk 3 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/projects/create.js]: @wtrocki tks for the review. However, the original command just works with the position as follows. ( the name in the desc was project-title either)

`````
Camilas-MacBook-Pro:fh-fhc cmacedo$ fhc projects create
(node) sys is deprecated. Use util instead.
fhc ERR! Invalid arguments
fhc ERR! Usage: 
fhc ERR! fhc projects [list] [<author-email>]
fhc ERR! fhc projects create <project-title> [<template-id>] --env=<environment>
fhc ERR! fhc projects update <project-id> <prop-name> <value>
fhc ERR! fhc projects read <project-id>
fhc ERR! fhc projects delete <project-id>
fhc ERR! fhc projects clone <project-id>
fh
````

**Comment by jasonmadigan** [comment]
@camilamacedo86 @wtrocki @kpiwko think `fh-uart` still uses fhc project create heavily, so we should also ensure this works for UART too

**Comment by camilamacedo86** [comment]
@jasonmadigan  add test here in the first comment with the --json which give issue for them in another PR. 
c/c @kpiwko

**Comment by kpiwko** [comment]
@trepel could you please check this pr with core-art and fh-uart to get idea how much backwards compat this could break?

fh-art is not used anymore.

**Comment by camilamacedo86** [comment]
@wtrocki Please, could you approve this one? You put that all is ok but you leave with not approved.

File: lib/cmd/fh3/projects/create.js (hunk 3 above)
Comment (wtrocki) [inline lib/cmd/fh3/projects/create.js]: Little bit misunderstanding:
Old command format was: `fhc projects create <project-title>`
Currently: `fhc projects create --title=<title>`

While I agree to keep it consistent let's not break previous implementations.
Am I getting it right?

**Comment by trepel** [comment]
@wtrocki @camilamacedo86  I am really sorry for my ignorance but I didn't understand what was the outcome of your discussion here. As @wtrocki  mentioned in one of his earlier comments this PR introduces breaking change. I am confused whether we want to merge this or not. If yes, this PR will definitely require another commit with just version bump. If not, Camila will need to update this PR so that the usage of fhc projects is not changed.

**Comment by trepel** [comment]
@wtrocki @camilamacedo86 I tried to use this PR with some of our test suites. Here some modifications were required. For instance in mbaas-art if using this PR's version of fhc I encountered `TypeError: fhc.projects is not a function`. This was due to this piece of code:  `fhc.projects({_: ['create', params.projectName, params.template], 'env': 'none'}`

I will try to provide more details tomorrow but it seems that if our customers are using fhc this way, this can be breaking change for them.

**Comment by camilamacedo86** [comment]
@trepel the customer are not using this project as dependence. So, they don't will face it. 
We need change the call the way that pass the values for as follows.
fhc.projects.create({params} 
I will create a JIRA for change the calls and do this PR. Thank you so much for your test. 

c/c @jasonmadigan

**Comment by trepel** [comment]
@camilamacedo86 Thanks for you PR. My question was more about customer usage. I don't know if customers use fhc as library or not so I can't tell whether this is OK to merge from customer perspective.

I any case I would consider adding another commit just for version bump. The need for updating internal test suites because of this change should not definitely be a blocker for merging.

**Comment by wtrocki** [comment]
> the customer are not using this project as dependence
Let's not make assumptions for the part outside our control. fh-fhc can be used as library and command line client. Our scripts using both options. Let's try to make improvements without breaking compatibility. For this particular example:

Old command format was: `fhc projects create <project-title>`
Currently: `fhc projects create --title=<title>`

**Comment by trepel** [comment]
> Old command format was: fhc projects create <project-title>
Currently: fhc projects create --title=<title>

That is no longer true @wtrocki , both ways can be used. Therefore if using fh-fhc as a command line tool, there is no breaking change. Breaking change can be encountered only if fh-fhc is used as a library.

**Comment by camilamacedo86** [comment]
@wtrocki @trepel Tks for all so far :-)

**Comment by camilamacedo86** [comment]
@trepel, 

We can't show ugly messages for users. The messages need be useful. 

After the change made into the PR[1], when you call the commands as dependecy and pass the parameter json the result will be the json of the request made. So you will be still able to check and validate your values in the projects to make tests or any one that want use as dependecy project. 

About the bump version: When it be bumped is generate a new version. We don't need do it now. We are doing the implementations and we can just bump the version in this project when we would like to generate a new production version. 

[1] - https://github.com/feedhenry/fh-fhc/pull/349

**Comment by trepel** [comment]
@camilamacedo86 Thanks for pointing me to the other PR, I will have a look at that PR and then combine  the two PRs and verify.

**Comment by trepel** [comment]
@camilamacedo86 As we discussed via bluejeans, for test automation there needs to be the way to get json output instead of human readable messages. You said that having the --json switch for this is a possible solution and I agree with that. You also said you implemented it in https://github.com/feedhenry/fh-fhc/pull/349. Unfortunately I can't see how that PR implements it for `fhc projects create` command .

I would suggest to add that --json switch (similarly as you did for list.js here) to other commands so that using this switch the commands will have the same json output as before. In such case users/customers using fh-fhc as a library (dependency) will just have to do simple update of their scripts. And I will be able to approve this PR.

**Comment by camilamacedo86** [comment]
@trepel is impossible do the same implementation made into  #349.for the projects before merge this one or that one. 

For merge a PR is required update with the master. So, when one of both me merged this will be done for the fh projects too.

**Comment by camilamacedo86** [comment]
@trepel , it is updated with the master and the required adjustments was made.
Also, I re-test all commands locally either. Please, feel free to test it.

**Comment by trepel** [comment]
@camilamacedo86 Hi Camila, I tested the PR and now it looks good. Feel free to rebase and merge.

**Comment by camilamacedo86** [comment]
@trepel  tks :-)

**Pull request merged**